### PR TITLE
Creating URI-Domain root path with DataTree special paths.

### PR DESCRIPTION
**Pull request by rahulrane50**
Problem : 
URI-domain root path is currently created in-memory with ZKDataBase APIs and this is making digest mismatch errors. Also the node is not correctly created and although listing "/zookeeper" shows it but creation of any znode under that path fails with missing znode error.

Fix : 
This could happen because we are using ZKDataBase in-memory API which could have digest mismatch and consistency issues. Right now zookeeper-server creates couple of special paths like "/zookeeper", "/zookeeper/config"., etc under root path. Just like we can create another uri-domain root path as special path if it's ZNodeGroupACL authentication provider.

Testing : 
Verified that znode is corrected created and persisted.
Verified transaction logs and snapshot for validity.
Verified path won't be created if not using ZNodeGroupACL provider.

### CODE REVIEW

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/DataTree.java [hunk 1]
```diff
@@ -366,6 +376,30 @@ public void addConfigNode() {
...
+        DataNode zookeeperZnode = nodes.get(procZookeeper);
+        if (zookeeperZnode != null) { // should always be the case
+            zookeeperZnode.addChild(uriDomainMapChildZookeeper);
+        } else {
+            assert false : "There's no /zookeeper znode - this should never happen.";
+        }
+
+        nodes.put(uriDomainMapZookeeper, new DataNode(new byte[0], -1L, new StatPersisted()));
+        try {
+            // Reconfig node is access controlled by default (ZOOKEEPER-2014).
```
Comment (abhilash1in) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/DataTree.java]: How is this comment relevant here?

Comment (narendly) [reply zookeeper-server/src/main/java/org/apache/zookeeper/server/DataTree.java]: +1

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/DataTree.java [hunk 2]
```diff
@@ -366,6 +376,30 @@ public void addConfigNode() {
...
+            setACL(uriDomainMapZookeeper, ZooDefs.Ids.OPEN_ACL_UNSAFE, -1);
+        } catch (KeeperException.NoNodeException e) {
+            assert false : "There's no " + uriDomainMapZookeeper + " znode - this should never happen.";
```
Comment (narendly) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/DataTree.java]: Is this check necessary? Consider just wrapping the whole block in a `try-catch` with a proper exception log?

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/DataTree.java [hunk 3]
```diff
@@ -366,6 +376,30 @@ public void addConfigNode() {
...
+    public void addUriDomainMapNode() {
+        if (!X509AuthenticationConfig.getInstance().isX509ZnodeGroupAclEnabled()) {
+            return;
+        }
+        DataNode zookeeperZnode = nodes.get(procZookeeper);
+        if (zookeeperZnode != null) { // should always be the case
+            zookeeperZnode.addChild(uriDomainMapChildZookeeper);
+        } else {
+            assert false : "There's no /zookeeper znode - this should never happen.";
+        }
```
Comment (narendly) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/DataTree.java]: A few comments:
1. In Java, we usually throw exceptions over `assert`s.
2. I'm not sure if that check is necessary here - like you said, `zookeeperZNode` should always exist? Then can we simplify things?

Comment (NealSun96) [reply zookeeper-server/src/main/java/org/apache/zookeeper/server/DataTree.java]: +1, if /zookeeper doesn't exist, we should let it fail.

**Review by NealSun96 — CHANGES_REQUESTED** [review]

File: zookeeper-server/src/main/java/org/apache/zookeeper/ZooDefs.java [hunk 4]
```diff
@@ -32,6 +32,10 @@ public class ZooDefs {
 
     public static final String ZOOKEEPER_NODE_SUBTREE = "/zookeeper/";
 
+    /** Root path for storing client URI-domain mappings for X509ZNodeGroupAclProvider. For more information
+     * about what info is stored under this path please see ZkClientUriDomainMappingHelper documentation. */
+    public static final String URI_DOMAIN_MAP_NODE = "/zookeeper/uri-domain-map";
```
Comment (mgao0) [inline zookeeper-server/src/main/java/org/apache/zookeeper/ZooDefs.java]: Let's not hard code it, this is one of the configs that we read in https://github.com/linkedin/zookeeper/blob/262b35635cb4fad0d70d70a2e75550fe845f9fd4/zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/X509AuthenticationConfig.java#L132. Or do you mean that we should strictly define this path, to make sure that this is a child node of /zookeeper?

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/DataTree.java [hunk 5]
```diff
@@ -146,6 +147,13 @@ public class DataTree {
     /** this will be the string thats stored as a child of /zookeeper */
     private static final String configChildZookeeper = configZookeeper.substring(procZookeeper.length() + 1);
 
+    /** Root path for storing client URI-domain mappings for X509ZNodeGroupAclProvider. For more information
+     * about what info is stored under this path please see ZkClientUriDomainMappingHelper documentation. */
+    private static final String uriDomainMapZookeeper = ZooDefs.URI_DOMAIN_MAP_NODE;
```
Comment (mgao0) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/DataTree.java]: Same as above, is there a reason to hard code it?

**Review by mgao0 — COMMENTED** [review]
Great work! Please modify the test in ZkClientUriDomainMappingHelperTest.testA_ZkClientUriDomainMappingHelper to make sure that we can create child nodes under this root path.

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/DataTree.java (hunk 3 above)
Comment (rahulrane50) [reply zookeeper-server/src/main/java/org/apache/zookeeper/server/DataTree.java]: In general i agree with your thoughts but here i was just trying to follow things just like addConfigNode method. I don't have full context on these checks but i'm fine with keeping it just like addConfigNode method or changing it completely.

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/DataTree.java (hunk 1 above)
Comment (rahulrane50) [reply zookeeper-server/src/main/java/org/apache/zookeeper/server/DataTree.java]: Same as above comment

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/DataTree.java (hunk 2 above)
Comment (rahulrane50) [reply zookeeper-server/src/main/java/org/apache/zookeeper/server/DataTree.java]: Same as above comment

**Review by rahulrane50 — COMMENTED** [review]
@narendly @NealSun96

File: zookeeper-server/src/main/java/org/apache/zookeeper/ZooDefs.java (hunk 4 above)
Comment (narendly) [reply zookeeper-server/src/main/java/org/apache/zookeeper/ZooDefs.java]: I think there's value in making this a nested path under `/zookeeper` because that way this will be exempt from digest calculations. Now, we might want to think about the fact that the URI-domain mapping path is configurable - should we modify our assumptions and make it always a sub-path of `/zookeeper`?

Comment (rahulrane50) [reply zookeeper-server/src/main/java/org/apache/zookeeper/ZooDefs.java]: After trying out few things looks like anyways nested paths under /zookeeper won't work in all cases so even if we make this configurable then it's only about the name of the directory but the actual path so to remove any confusion i made it fixed let me know what you guys think.

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/DataTree.java [hunk 6]
```diff
@@ -377,7 +407,8 @@ boolean isSpecialPath(String path) {
         return rootZookeeper.equals(path)
                || procZookeeper.equals(path)
                || quotaZookeeper.equals(path)
-               || configZookeeper.equals(path);
+               || configZookeeper.equals(path)
+               || (X509AuthenticationConfig.getInstance().isX509ZnodeGroupAclEnabled() && uriDomainMapZookeeper.equals(path));
```
Comment (narendly) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/DataTree.java:411]: Nit: `&&` takes precedence over `||`, so the parenthesis pair is probably not needed?

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/X509AuthenticationConfig.java [hunk 7]
```diff
@@ -124,20 +124,19 @@ public static X509AuthenticationConfig getInstance() {
...
   private String znodeGroupAclSuperUserId;
   private String znodeGroupAclCrossDomainAccessDomainNameStr;
   private String znodeGroupAclOpenReadAccessPathPrefixStr;
   private String znodeGroupAclServerDedicatedDomain;
-  private String znodeGroupAclClientUriDomainMappingRootPath;
+
+  // The root path for client URI - domain mapping that stored in zk data tree. Refer to
+  // {@link org.apache.zookeeper.server.auth.znode.groupacl.ZkClientUriDomainMappingHelper} for
+  // details about client URI - domain mapping. The root path is fixed and kept under "/zookeeper" to
+  // get auto-created during server bootup process.
```
Comment (narendly) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/X509AuthenticationConfig.java]: right on!

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/znode/groupacl/ZkClientUriDomainMappingHelper.java [hunk 8]
```diff
@@ -36,7 +42,7 @@
 import org.slf4j.LoggerFactory;
 
 /**
- * An implementation of ClientUriDomainMappingHelper that stores the mapping inside the ZK server
+ * An implementation of ZkClientUriDomainMappingHelper that stores the mapping inside the ZK server
```
Comment (narendly) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/znode/groupacl/ZkClientUriDomainMappingHelper.java]: Since `ClientUriDomainMappingHelper` is the interface this class `implements`, I'm not sure why we need to change this?

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/znode/groupacl/ZkClientUriDomainMappingHelper.java [hunk 9]
```diff
@@ -45,7 +51,7 @@
  *
  * The following illustrates the ZNode hierarchy:
  * . (root)
- * └── _CLIENT_URI_DOMAIN_MAPPING (mapping root path)
+ * └── /zookeeper/uri-domain-map (mapping root path)
```
Comment (narendly) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/znode/groupacl/ZkClientUriDomainMappingHelper.java:54]: Should we reflect here that `uri-domain-map` is actually a configurable (and therefore variable) string?

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/DataTree.java [hunk 10]
```diff
@@ -384,19 +384,15 @@ public void addUriDomainMapNode() {
...
 
-        nodes.put(uriDomainMapZookeeper, new DataNode(new byte[0], -1L, new StatPersisted()));
-        try {
-            // Reconfig node is access controlled by default (ZOOKEEPER-2014).
+            nodes.put(uriDomainMapZookeeper, new DataNode(new byte[0], -1L, new StatPersisted()));
             setACL(uriDomainMapZookeeper, ZooDefs.Ids.OPEN_ACL_UNSAFE, -1);
-        } catch (KeeperException.NoNodeException e) {
-            assert false : "There's no " + uriDomainMapZookeeper + " znode - this should never happen.";
+            LOG.info("Successfully created client URI-domain root path :{}", uriDomainMapZookeeper);
+        } catch (Exception e) {
```
Comment (mgao0) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/DataTree.java:394]: Should we fail here if the root path creation is failed?

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/znode/groupacl/ZkClientUriDomainMappingHelper.java (hunk 9 above)
Comment (rahulrane50) [reply zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/znode/groupacl/ZkClientUriDomainMappingHelper.java:54]: Sorry i didn't get you. It's not configurable anymore after my change right? @narendly

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/DataTree.java (hunk 6 above)
Comment (rahulrane50) [reply zookeeper-server/src/main/java/org/apache/zookeeper/server/DataTree.java:411]: True it's not needed logically but just for the easy code reading purpose i added it. :)

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/DataTree.java (hunk 10 above)
Comment (rahulrane50) [reply zookeeper-server/src/main/java/org/apache/zookeeper/server/DataTree.java:394]: If we fail here itself then customer won't even be able to use any other authentication provider or plaintext port. Hence i just added log here but raising an exception in ZkClientUriDomainMappingHelper constructor if root path is not present. Does it make sense? @mgao0

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/X509AuthenticationConfig.java [hunk 11]
```diff
@@ -124,20 +124,19 @@ public static X509AuthenticationConfig getInstance() {
...
+  private static final String znodeGroupAclClientUriDomainMappingRootPath = "/zookeeper/uri-domain-map";
```
Comment (narendly) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/X509AuthenticationConfig.java]: Now this is a constant, so let's follow the naming convention (all caps). See [this](https://stackoverflow.com/questions/7259687/java-naming-convention-for-static-final-variables)

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/znode/groupacl/ZkClientUriDomainMappingHelper.java (hunk 9 above)
Comment (narendly) [reply zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/znode/groupacl/ZkClientUriDomainMappingHelper.java:54]: Thanks for pointing it out - I missed it. If that's the case, I think it's good as is! Resolving.

**Review by narendly — APPROVED** [review]

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/DataTree.java (hunk 10 above)
Comment (mgao0) [reply zookeeper-server/src/main/java/org/apache/zookeeper/server/DataTree.java:394]: Make sense. Thank you!

**Review by mgao0 — APPROVED** [review]
